todo_app/app.py

```diff
@@ -1,49 +1,127 @@
-from flask import Flask, render_template, request, redirect, url_for
+from flask import Flask, render_template, request, redirect, url_for, abort
+from flask_login import LoginManager, login_required
+import flask_login
+from flask_login.mixins import UserMixin
+from oauthlib.oauth2 import WebApplicationClient
+from oauthlib.oauth2.rfc6749.grant_types import client_credentials
 import Mongo_items as mongo
 import viewmodel as vm
+import os, requests, json
+from functools import wraps
 
 def create_app():
     app = Flask(__name__)
     app.config.from_object('flask_config.Config')
 
+    login_manager = LoginManager()
+
+    @login_manager.unauthorized_handler
+    def unauthenticated():
+        client = WebApplicationClient(os.getenv("CLIENT_ID"))
+        authredirect = client.prepare_request_uri("https://github.com/login/oauth/authorize", state="saxasdfaa")
+        return redirect(authredirect)
+
+    @login_manager.user_loader
+    def load_user(user_id):
+        user = User(user_id)
+        return user
+
+    login_manager.init_app(app)
+
+    class User(UserMixin):
+        def __init__(self, userid):
+            self.user_id = userid
+            self.id = userid
+            if self.id == "69510597" :
+                self.role = 'reader'
+            else :
+                self.role = 'writer'
+
+    def checkrole(f):
+        @wraps(f)
+        def decorated_function(*args, **kwargs):
+
+            user = load_user(flask_login.current_user.id)
+            if user.role == 'reader':
+                abort(403)
+            else:
+                return f(*args, **kwargs)
+        return decorated_function
+
     @app.route('/')
+    @login_required
     def index():
         items = mongo.get_items_mongo()
         items=sorted(items, key=lambda k: k.status, reverse=True)
         item_view_model = vm.ViewModel(items)
+        user = load_user(flask_login.current_user.id)
+        if user.role == 'reader':
```

Comment:
(Same issue here with this erroring for anonymous users)

Note that this pattern requires a bit more upkeep than if you passed through the role to the template, and determined section by section what can/can't be shown, since changes may well need to be applied to two separate templates here

Reply:
good point - I had thought I could do it in the template but hadn't explored that further so I've got that working now